ckan_meta_tester/ckan_meta_tester.py

```diff
@@ -0,0 +1,258 @@
+import re
+from os import environ
+from shutil import copy
+import logging
+from git import Repo, Commit
+from subprocess import run, Popen, PIPE, STDOUT
+from pathlib import Path
+from importlib.resources import read_text
+from string import Template
+from exitstatus import ExitStatus
+from typing import Optional, Iterable, Set, List, Any, Dict, Tuple
+
+from .ckan_install import CkanInstall
+from .game_version import GameVersion
+from .dummy_game_instance import DummyGameInstance
+from .log_group import LogGroup
+
+
+class CkanMetaTester:
+
+    # Location of the netkan.exe and ckan.exe files in the container
+    BIN_PATH    = Path('/usr/local/bin')
+    NETKAN_PATH = BIN_PATH.joinpath('netkan.exe')
+    CKAN_PATH   = BIN_PATH.joinpath('ckan.exe')
+
+    INFLATED_PATH = Path('/ckans')
+    CACHE_PATH    = Path('/cache')
+    REPO_PATH     = Path('/repo')
+    TINY_REPO     = REPO_PATH.joinpath('metadata.tar.gz')
+
+    CKAN_INSTALL_TEMPLATE = Template(read_text(
+        'ckan_meta_tester', 'ckan_install_template.txt'))
+
+    PR_BODY_COMPAT_PATTERN = re.compile('ckan compat add((?: [0-9.]+)+)')
+
+    REF_ENV_VARS = [
+        'GITHUB_PR_BASE_SHA',
+        'GITHUB_EVENT_BEFORE'
+    ]
+
+    def __init__(self) -> None:
+        self.source_to_ckan: Dict[Path, Path] = {}
+        self.failed = False
+
+    def test_metadata(self, source: str = 'netkans', pr_body: str = '', github_token: Optional[str] = None) -> bool:
+
+        logging.debug('Starting metadata test')
+        logging.debug('Builds: %s', [str(v) for v in CkanInstall.KNOWN_VERSIONS])
+
+        # Escape hatch in case author replaces a download after a previous success
+        # (which will save it to the persistent cache)
+        overwrite_cache = ('#overwrite_cache' in pr_body)
+        logging.debug('overwrite_cache: %s', overwrite_cache)
+
+        for file in self.files_to_test(source):
+            if not self.test_file(file, overwrite_cache, github_token):
+                logging.error('Test of %s failed!', file)
+                self.failed = True
+        if self.failed:
+            return False
+
+        if len(self.source_to_ckan) == 0:
+            logging.info('No .ckans found, done.')
+            return True
+
+        # Make secondary repo file with our generated .ckans
+        run(['tar', 'czf', self.TINY_REPO, '-C', self.INFLATED_PATH, '.'])
+
+        for orig_file, file in self.source_to_ckan.items():
+            if not self.install_ckan(file, orig_file, pr_body):
+                logging.error('Install of %s failed!', file)
+                self.failed = True
+
+        return not self.failed
+
+    def test_file(self, file: Path, overwrite_cache: bool, github_token: Optional[str] = None) -> bool:
+        logging.debug('Attempting jsonlint for %s', file)
+        if not self.run_for_file(
+            file, ['jsonlint-php', file], full_output_as_error=True):
```

Comment:
There are a number of Python tools we can use for linting JSON. But we might want to develop a wrapper + tests as they often don't print a super nice output.

Reply:
Yeah, I was going for a like-for-like solution to start out; the old script called the external `jsonlint` tool, and this was the closest replacement I could find in ubuntu-latest. If there's a good Python lib I'd be all for switching it.